sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -132,11 +142,14 @@ object ConstantPropagation extends Rule[LogicalPlan] with PredicateHelper {
         (newSelf, Seq.empty)
       case n: Not =>
         // Ignore the EqualityPredicates from children since they are only propagated through And.
-        val (newChild, _) = traverse(n.child, replaceChildren = true)
+        val (newChild, _) = traverse(n.child, replaceChildren = true, nullIsFalse = false)
         (newChild.map(Not), Seq.empty)
       case _ => (None, Seq.empty)
     }
 
+  private def safeToReplace(ar: AttributeReference, nullIsFalse: Boolean) =
```

Comment:
Ah, I see. Good catch. Can you leave some comments (w/ an example query) here for easy-to-understand (about what this condition intend to avoid)?

Reply:
Ok, added. Let me know if it needs rewording.